### ISSUE TITLE
Repair checkpoint for checkpoint corruption and not the WAL

### ISSUE THREAD

**Issue by codesome** [prometheus/prometheus#7530]
When we try to replay a checkpoint here 
https://github.com/prometheus/prometheus/blob/30505a202a4c33ceeb10bfb3ba01e371a2a10906/tsdb/head.go#L678-L680

we can encounter a `wal.CorruptionErr`. But we are repairing only the WAL irrespective of whether the corruption was from WAL or checkpoint:
https://github.com/prometheus/prometheus/blob/30505a202a4c33ceeb10bfb3ba01e371a2a10906/tsdb/db.go#L632-L638

This can lead to wrong repairs as segment numbers inside checkpoint and WAL are not related. We should be repairing checkpoints for corruption in the checkpoint.

Currently, with checkpoint errors, we lose all the data from WAL. But if we could repair checkpoint, we can replay samples for at least some portion of the series. (Irrespective of this, WAL should not be repaired for checkpoint corruption)

**Comment by brian-brazil** [comment]
> But if we could repair checkpoint, we can replay samples for at least some portion of the series.

Only some series isn't of much use, for the same reasons a partial scrape is worse than a failed scrape - and would effecitvely bring back the unknown series errors. If there's corruption we need to drop all data from the corruption onwards, which in the case of the checkpoint means dropping the entire checkpoint.

**Comment by codesome** [comment]
Do you mean to say drop both checkpoint and also the WAL? (btw, in that case, we should also remove the data from the memory)

**Comment by brian-brazil** [comment]
We can't use the WAL without the checkpoint (if there is a checkpoint), so if there's checkpoint corruption we should be starting with nothing in the head.

**Comment by showuon** [comment]
Will work on this. Thanks.

**Comment by Harkishen-Singh** [comment]
@showuon are you still working on this?

**Comment by codesome** [comment]
Hold on before working on this. https://github.com/prometheus/prometheus/issues/7530#issuecomment-655393626 makes sense and need to re-think this. If checkpoint is corrupt, replaying WAL for partial series might not be useful/meaningful.

**Comment by codesome** [comment]
Hold on before working on this. https://github.com/prometheus/prometheus/issues/7530#issuecomment-655393626 makes sense and need to re-think this. If checkpoint is corrupt, replaying WAL for partial series might not be useful/meaningful.

**Comment by showuon** [comment]
Sorry. No, go ahead to take over it. Thanks.

**Comment by Harkishen-Singh** [comment]
Definitely! @codesome please let me know once you have had thought about the approach in the comment by Brian.

**Comment by codesome** [comment]
In Prometheus Storage Working Group call we wen't over these 3 options of how we want to handle this

1. Drop everything from Head since there will be partial series. This means if you have any corruption in checkpoint, you lose all data that was there in the Head.
2. Repair the checkpoint and continue with the remaining WAL replay. This means you will lose partial number of series (for the series records that were present in the checkpoint from after the corruption), and there will possibly be some gap in the restored series (again, for the missing samples record in the corrupted section of checkpoint), but the remaining samples for those series from the WAL segments will be restored. 
3. Repair the checkpoint and restore the partial number of series as option 2, but don't replay remaining of the WAL. Hence you lose a bunch of samples for the restored series that's in the WAL segments.

The proposal here is option 2 which was in the original issue description (I have updated it to reflect this).

Here the option 3 does not look attractive in front of option 2, and we do not want to drop all the data. While partial series could return wrong results for some queries (but the operator should already be notified about corruption hence expect this), we cannot rule out the use cases where partial series is fine. 

@bwplotka @brancz can you please confirm the above?

**Comment by Harkishen-Singh** [comment]
That is good. Will go with the fix once bartek  confirms.

**Comment by roidelapluie** [comment]
I do not think we should go with option 2. partial data is impossible to work with.

**Comment by codesome** [comment]
Then that sounds like option 1. Also, checkpoint corruptions should be rare, it will only happy if disk is faulty and not on Prometheus crashes.

**Comment by codesome** [comment]
To not block this any further, let's go with the simpler option 1, and we can revisit this later if that is causing any problems.

**Comment by Harkishen-Singh** [comment]
Then I will go with option 1. Thank you!

If someone has any other concerns, please comment here.

**Comment by krajorama** [comment]
Hello from the bug scrub: there seems to be open PRs that address the issue, keeping open.

**Comment by beorn7** [comment]
Hello again from the bug-scrub! 

We have closed the following PRs that tried to tackle this because there was no movement on them for a long time and no volunteer to take them over. We'll note them here so future takers of this issue can take inspiration from them:
- https://github.com/prometheus/prometheus/pull/9148
- https://github.com/prometheus/prometheus/pull/8205
